## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [243]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [244]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [245]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [246]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [247]:
print(f"Sanity Check\n\n"
      f"Order Years: {df["order_year"].unique()}\n"
      f"Category {df["category"].unique()}\n"
      f"Subcategory {df["subcategory"].unique()}\n"
      f"\nNumber of nulls\n"
      f"{df.isnull().sum()}"
)

Sanity Check

Order Years: [2023 2022]
Category ['decor' 'consumables' 'software' 'furniture' 'electronics']
Subcategory ['wall art' 'pens' 'project management' 'misc' 'plants' 'kitchen supplies'
 'paper products' 'seating' 'computers' 'stands' 'tables'
 'bathroom supplies' 'audio visual' 'accounting' 'logistics' 'telecom'
 'desks' 'cabling']

Number of nulls
first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64


In [248]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [249]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df["category"].value_counts().head(3).index[0]
df["subcategory"].loc[df["category"] == top_category].value_counts()

subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [250]:
# Which five clients had the most entries in the data?
top_clients_df = df["client_id"].value_counts().head()
top_clients_df

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [251]:
# Store the client ids of those top 5 clients in a list.
top_clients = top_clients_df.index.to_list()
top_clients

[33615, 66037, 46820, 38378, 24741]

In [252]:
# How many total units (the qty column) did the client with the most entries order order?
df["qty"].loc[df["client_id"] == top_clients[0]].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [253]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]
df[["unit_price", "qty", "line_subtotal"]].head(3)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28

In [254]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calc_shipping(weight):
    rate = 10.00
    if weight > 50:
        rate = 7.00
    return format(rate*weight, ".2f")

df["total_weight"] = df["unit_weight"] * df["qty"]
df["shipping_price"] = df["total_weight"].apply(calc_shipping).astype(float)
df[["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]].head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [255]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   first           54639 non-null  object 
 1   last            54639 non-null  object 
 2   job             54639 non-null  object 
 3   phone           54639 non-null  object 
 4   email           54639 non-null  object 
 5   client_id       54639 non-null  int64  
 6   order_id        54639 non-null  int64  
 7   order_date      54639 non-null  object 
 8   order_week      54639 non-null  int64  
 9   order_year      54639 non-null  int64  
 10  item_id         54639 non-null  object 
 11  category        54639 non-null  object 
 12  subcategory     54639 non-null  object 
 13  unit_price      54639 non-null  float64
 14  unit_cost       54639 non-null  float64
 15  unit_weight     54639 non-null  float64
 16  qty             54639 non-null  int64  
 17  line_number     54639 non-null 

In [256]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["line_price"] = ((df["line_subtotal"] + df["shipping_price"]) * 1.0925).round(2)
df[["line_subtotal", "shipping_price", "line_price"]].head(3)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [257]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"] = df["unit_cost"] * df["qty"] + df["shipping_price"]
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [258]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df["line_price"] - df["line_cost"]
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [259]:
# Check your work using the totals above
def total_order_price(order_id):
    order = df.loc[df["order_id"] == order_id]
    return order["line_price"].sum()

order_id = 2742071
expected = "$152,811.89"
print(f"Order ID {order_id}")
print(f"${total_order_price(order_id):,.2f}")
print(f"{expected}\n")

order_id = 2173913 
expected = "$162,388.71"
print(f"Order ID {order_id}")
print(f"${total_order_price(order_id):,.2f}")
print(f"{expected}\n")


order_id = 6128929 
expected = "$923,441.25"
print(f"Order ID {order_id}")
print(f"${total_order_price(order_id):,.2f}")
print(f"{expected}\n")


Order ID 2742071
$152,811.89
$152,811.89

Order ID 2173913
$162,388.71
$162,388.71

Order ID 6128929
$923,441.25
$923,441.25



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [260]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client_id in top_clients:
    orders = df.loc[df["client_id"] == client_id, ["line_price"]]
    total = orders["line_price"].sum()
    print(f"Client {client_id} in total spent ${total:,.2f}")


Client 33615 in total spent $8,377,308.52
Client 66037 in total spent $10,259,514.79
Client 46820 in total spent $9,743,794.36
Client 38378 in total spent $12,906,550.87
Client 24741 in total spent $82,268,892.02


In [261]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = pd.DataFrame(columns=["client_id", "qty", "shipping_price", "line_price", "line_profit"])

for client_id in top_clients:
    orders = df.loc[df["client_id"] == client_id, ["qty", "shipping_price", "line_price", "line_profit"]]
    total_units_purchased = orders["qty"].sum()
    total_shipping_price = orders["shipping_price"].sum()
    total_revenue = orders["line_price"].sum()
    total_profit = orders["line_profit"].sum()

    summary_df = summary_df._append({ # Blackbox
        "client_id": int(client_id),
        "total_units_purchased": int(total_units_purchased),
        "total_shipping_price": total_shipping_price,
        "total_revenue": total_revenue,
        "total_profit": total_profit},
        ignore_index=True)

#     print(f"For Client {client_id}:\n"
#           f"Total Units Purchased: {total_units_purchased} units\n"
#           f"Total Shipping Price: ${total_shipping_price:,.2f}\n"      
#           f"Total Revenue: ${total_revenue:,.2f}\n"      
#           f"Total Profit: ${total_profit:,.2f}\n"      
#     )
summary_df["client_id"] = summary_df["client_id"].astype(int)
summary_df["qty"] = summary_df["qty"].astype(int)
summary_df

C:\Users\fsgin\AppData\Local\Temp\ipykernel_23980\3476581326.py:12: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  summary_df = summary_df._append({ # Blackbox


client_id     qty  shipping_price   line_price  line_profit
0      33615   64313      1828984.89   8377308.52   2201994.61
1      66037   43018      1395151.85  10259514.79   3255031.81
2      46820   75768      1601448.84   9743794.36   2736602.72
3      38378   73667      3429455.40  12906550.87   3271829.89
4      24741  239862      5126448.37  82268892.02  36579992.31

In [262]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.rename(columns={
    "client_id": "Client ID",
    "total_units_purchased": "Units Purchased",
    "total_shipping_price": "Total Shipping Price",
    "total_revenue": "Revenue Generated",
    "total_profit": "Proft Generated",
})

# Define the money columns. 
money_columns = ["Units Purchased", "Total Shipping Price", "Revenue Generated", "Proft Generated"]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(money):
    return format(money / 1e6, ",.2f")

# Apply the currency_format_millions function to only the money columns. 

# Rename the columns to reflect the change in the money format. 


In [263]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
